### DISCUSSION TITLE
Support TypeBox validation library

### DISCUSSION THREAD

**Discussion by peguerosdc** [next-safe-action/next-safe-action#225] [Ideas]
## Context

Hi! I wanted to try `next-safe-action@7.4.3`, but all my validations depend on [sinclairzx81/typebox](https://github.com/sinclairzx81/typebox) because `zod` has historically given me performance issues, so I tried to create a `SafeActionClient` for this schema validation library by mimicking what I found in [next-safe-action/src/adapters](https://github.com/TheEdoRan/next-safe-action/tree/main/packages/next-safe-action/src/adapters):

```typescript
import { createSafeActionClient } from "next-safe-action";
import {
  IfInstalled,
  ValidationAdapter,
} from "next-safe-action/adapters/types";
import { type TSchema } from "@sinclair/typebox";
import { Value } from "@sinclair/typebox/value";

class TypeboxAdapter implements ValidationAdapter {
  async validate<S extends IfInstalled<TSchema>>(schema: S, data: unknown) {
    try {
      const result = Value.Decode(schema, data);
      return {
        success: true,
        data: result,
      } as const;
    } catch (e) {
      const errors = Array.from(Value.Errors(schema, data));
      return {
        success: false,
        issues: errors.map(({ message, path }) => ({
          message,
          path: [path],
        })),
      } as const;
    }
  }
}

const typeboxAdapter = () => {
  return new TypeboxAdapter();
};

export const actionClient = createSafeActionClient({
  validationAdapter: typeboxAdapter(),
});
```

## The Problem

I can't use the client without typescript complaining:

```typescript
import { Type } from "@/lib/typebox";

export const myAction = actionClient
  .schema(Type.Object({a: Type.String()}))
  .action(async (data) => { ... });
/* Argument of type 'TObject<{ a: TString; }>' is not assignable to parameter of type 'ZodType<any, ZodTypeDef, any> | ((prevSchema: undefined) => Promise<ZodType<any, ZodTypeDef, any>>)'.
 */
```

## Feature Request

I think the problem is that the `Schema`'s type definition [is coupled to the available libraries](https://github.com/TheEdoRan/next-safe-action/blob/d2ebfc3c49dee7d62ce5824c27375d8e7985680f/packages/next-safe-action/src/adapters/types.ts#L9-L13).

If you think it would be a good contribution, I can add `typebox` with some guidance to make the library fit my use-case for now, being aware that it would eventually need to be improved.

From what I saw, I think I'll need to create an `adapter` file with my code above and change the `adapter/types.ts` to make use of:

```typescript
import { Static } from "@/lib/typebox";

type Infer = Static<typeof schema>;
```

in both `Infer` and `InferIn`, as well as add the `IfInstalled<TSchema>` to the `Schema` type.

What do you think?

**Comment by TheEdoRan** [comment]
TypeBox support will be available in the library when #228 gets merged.

**Comment by TheEdoRan** [comment]
Support included in [v7.5.0](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.5.0)
